objects_delete/delete_temp_objects_job.py

```diff
@@ -0,0 +1,33 @@
+from time import time
+from gen3 import object, metadata, auth
+
+auth = auth.Gen3Auth(refresh_file="/mnt/api-profile-credentials.json")
+mds_handle = metadata.Gen3Metadata(auth)
+
+print("Initializing delete_temp_objects_job...")
+LIMIT_SIZE = 2
+offset_position = 0
+response_dict = mds_handle.query(
+    query="date_to_delete=*", return_full_metadata=True, limit=LIMIT_SIZE
+)
+if type(response_dict) is dict:
+    guid_list = list(response_dict.values())
+    while len(response_dict) == LIMIT_SIZE:
+        offset_position += LIMIT_SIZE
+        response_dict = mds_handle.query(
+            query="date_to_delete=*",
+            return_full_metadata=True,
+            limit=LIMIT_SIZE,
+            offset=offset_position,
+        )
+        if response_dict:
+            guid_list += list(response_dict.values())
+    delete_list = [obj["guid"] for obj in guid_list if obj["date_to_delete"] < time()]
+    object_api = object.Gen3Object(auth)
+    for guid in delete_list:
```

Comment:
You are keeping a lot of data in memory by constructing the whole `guid_list` first, and by storing all the metadata instead of only what you need (guid and date_to_delete). This could become a problem if there are a lot of records in MDS.
You are also going through the list twice, to construct `delete_list` and to delete, this could have an impact on performance if there are a lot of records.

To address both you could do something like this:
```
offset = 0
while:
    records = query()
    for r in records:
        if r.date < now:
            delete(r)
    offset += limit - (number of deleted records)
```

Reply:
Is it safe to assume that the order of all the records would always be in the same sequence? Somehow I automatically assumed the order wouldn't be reliable when I start deleting while querying. 💭 

If the order doesn't change, I can surely implement your suggestion, It surely does give better performance.

Reply:
yeah i was thinking about that - although the order doesn't change in most cases, technically there is no guarantee that it won't change, unless we specify "order by". I thought we could order by GUID in the MDS query (it might affect performance a little bit?), but if this runs at the same time as a data ingestion it will throw off our logic. I like the solution in indexd better ([sorting by updated_date](https://github.com/uc-cdis/indexd/blob/845256964a5a0bdd21da1aa6335fe9fb037318cc/indexd/index/drivers/alchemy.py#L490-L496)), unfortunately we don't store the date in MDS. In the end I figured making this assumption is fine here, but if you disagree we can keep querying everything before deleting, and just make a change so we don't go through the list twice to construct `delete_list` and to delete

Reply:
your code that iterates through the list twice:
```
    delete_list = [obj["guid"] for obj in guid_list if obj["date_to_delete"] < time()]
    for guid in delete_list:
        # delete(guid)
```
can be replaced with:
```
    for obj in guid_list:
        if obj["date_to_delete"] > time():
            continue
        # delete(obj["guid"])
```

Reply:
Ohh! You mean without the pagination? Got it.

Reply:
Oh wait, I think I got it wrong. You still want the `LIMIT` and `offset`, you just wanted to replace the list comprehension. You got it. (for sure this time :P)